pkg/services/ngalert/api/util_test.go

```diff
@@ -163,14 +163,15 @@ func Test_containsProvisionedAlerts(t *testing.T) {
 	t.Run("should return true if at least one rule is provisioned", func(t *testing.T) {
 		_, rules := models2.GenerateUniqueAlertRules(rand.Intn(4)+2, models2.AlertRuleGen())
 		provenance := map[string]models2.Provenance{
-			rules[rand.Intn(len(rules)-1)].UID: []models2.Provenance{models2.ProvenanceAPI, models2.ProvenanceFile}[rand.Intn(2)],
+			rules[rand.Intn(len(rules))].UID: []models2.Provenance{models2.ProvenanceAPI, models2.ProvenanceFile}[rand.Intn(2)],
 		}
 		require.Truef(t, containsProvisionedAlerts(provenance, rules), "the group of rules is expected to be considered as provisioned but it isn't. Provenances: %v", provenance)
 	})
 	t.Run("should return false if map does not contain or has ProvenanceNone", func(t *testing.T) {
 		_, rules := models2.GenerateUniqueAlertRules(rand.Intn(5)+1, models2.AlertRuleGen())
 		provenance := make(map[string]models2.Provenance)
-		for i := 0; i < rand.Intn(len(rules)); i++ {
+		numProvenanceNone := rand.Intn(len(rules))
```

Comment:
Can we just remove this randomness? It makes this test an absolute nightmare to read. Is it trying to test when a subset of the alerts are ProvenanceNone? What happens when none of them are? All of them are? This took me multiple read-throughs to understand what property it's trying to assert about the system.

Also, it appears from the name, it's not testing the right thing, intermittently.

The test's name suggests that it's intentionally trying to test when a few of the rules have ProvenanceNone, and others have some provenance. But, if the `rand.Intn()` call is zero, none of them have provenance, so it's not really testing anything! This could let an actual through the CI, a small fraction of the time. Let's just drop the randomness entirely.

Reply:
This test tests exactly what is declared: that the method returns false when the provenance map either does not contain any UID of the provided rules or all rules have provenanceNone in the map.

Reply:
I asked around and the possibility of the value being zero is intentional. I confirmed this by hardcoding a `0` and running the test.

I agree on the legibility thing, it took me a while to figure out what this test was about. This could be separated in two different tests:
1. No rules with provisioning info.
3. Some rules w/o info, some rules with `ProvenanceNone`.

This way we'd eliminate the randomness while having the same scenarios being tested.

Reply:
+1 for different tests